python/parquet_tools/common/schema.py

```diff
@@ -0,0 +1,700 @@
+                Schema._add_col(x[1])
+
+    # *************************************************************************
+    # *                        _ r e v e r t 2 n u l l                        *
+    # *************************************************************************
+    @classmethod
+    def _revert2null(cls, df, nilval):
+        """Revert integer fields that originally cotained null to be so.
+
+        Parameters
+        ----------
+        df : 'PANDAS dataframe'
+            This is the dataframe which is to be manipulated into submission.
+        nilval : 'string'
+            This is the sequence that defines a "null" value.
+
+        Raises
+        ------
+        FatalError
+        """
+
+        # If there are no null values at all, simply return.
+        #
+        if not Schema.colNVChk:
+            return None
+
+        # For each column that may have a null value, convert it to a string.
+        # Then replace it's value with a null sequence if it should contain it.
+        # Note that we need to get around flake8 as pandas only allows
+        # comparisons to values and we need to base execution on a boolean.
+        #
+        trueval = True
+        for cnum in Schema.colNVChk:
+            cname = Schema.colNames[cnum]
+            try:
+                df[cname] = df[cname].apply(str)
+            except Exception as ex:
+                raise FatalError(ex, 'Unable to convert integer column',
+                                 cname, 'to string')
+            cisnil = cname + Schema._nilSuffx
+            df.loc[df[cisnil] == trueval, cname] = nilval
+
+    # *************************************************************************
+    # *                           n o r m a l i z e                           *
+    # *************************************************************************
+    @classmethod
+    def normalize(cls, df, nilval):
+        """Normalize a PANDAS dataframe to correspond to the original schema.
+
+        Parameters
+        ----------
+        df : 'pandas dataframe'
+            The pandas dataframe to be normalized.
+        nilval : 'string'
+            This is the sequence that defines a "null" value.
+
+        Notes
+        - Normalization is defined as: a) null values are returned to integer
+          that could have contained null values, and b) floating point values
+          that represent fixed point data types are rounded to the original
+          precision.
+        """
+
+        # If we had any integer column that may have contained null values,
+        # make sure that rows in which null values occured are present.
+        #
+        if Schema.colNVChk:
+            Schema._revert2null(df, nilval)
+
+        # If we had any fixed point types are are being represented as floats,
+        # format the floats using the original precision.
+        #
+        for cname, Schema._colFmtfp in Schema._colFPFmt:
```

Comment:
so at this point the variable names become impossible to understand. do they have to be so tersely abbreviated? Can they be expanded into real english?